test/generators/gen_link/gen_link.ml

```diff
@@ -0,0 +1,214 @@
+      List [ Atom "target"; Atom ("page-" ^ basename ^ ".odocl") ];
+      List [ Atom "deps"; Atom ("page-" ^ basename ^ ".odoc") ];
+      List
+        [
+          Atom "action";
+          List
+            [
+              Atom "run";
+              Atom "odoc";
+              Atom "link";
+              Atom "-o";
+              Atom "%{target}";
+              Atom "%{deps}";
+            ];
+        ];
+    ]
+
+let mld_file_rule basename source_file =
+  let open Printf in
+  printf "%s\n\n" (to_string_hum (mld_odoc_target_rule basename source_file));
+  printf "%s\n\n" (to_string_hum (mld_odocl_target_rule basename))
+
+let die s =
+  prerr_endline s;
+  exit 1
+
+let path () f = Filename.quote (Fpath.to_string f)
+
+let ext' () f = Filename.quote (Fpath.get_ext f)
+
+let gen_rule_for_source_file source_file f =
+  let open Fpath in
+  let ext = get_ext f in
+  let basename = rem_ext f in
+  match ext with
+  | ".ml" -> file_rule (to_string basename) source_file ".cmt"
+  | ".mli" -> file_rule (to_string basename) source_file ".cmti"
+  | ".mld" -> mld_file_rule (to_string basename) source_file
+  | _ ->
+      die
+        (sprintf
+           "Don't know what to do with %a because of unrecognized %a extension."
+           path f ext' f)
+
+let odocls filenames =
+  filenames
+  |> List.map (fun f ->
+         let open Fpath in
+         if get_ext f = ".mld" then
+           sprintf "%%{dep:../page-%s.odocl}" (to_string (rem_ext f))
+         else sprintf "%%{dep:../%s.odocl}" (to_string (rem_ext f)))
+  |> List.map (fun f -> Atom f)
+
+let gen_rule_html odocls' =
+  List
+    [
+      Atom "subdir";
+      Atom "html";
+      List
+        [
+          Atom "rule";
+          List
+            [
+              Atom "with-stdout-to";
+              Atom "html.dune.inc.gen";
+              List
+                [
+                  Atom "pipe-stdout";
+                  List
+                    (Atom "run"
+                     :: Atom "gen_html/gen_html.exe" :: odocls odocls');
+                  List [ Atom "run"; Atom "dune"; Atom "format-dune-file" ];
+                ];
+            ];
+        ];
+    ]
+
+let diff_rule_html =
+  List
+    [
+      Atom "rule";
+      List [ Atom "alias"; Atom "runtest" ];
+      List
+        [
+          Atom "action";
+          List
+            [
+              Atom "diff";
+              Atom "html/html.dune.inc";
+              Atom "html/html.dune.inc.gen";
+            ];
+        ];
+    ]
+
+let gen_html_rule filenames =
+  printf "%s\n\n" (to_string_hum (gen_rule_html filenames));
+  printf "%s\n\n" (to_string_hum diff_rule_html)
+
+let read_file_from_dir dir =
+  let filenames = Sys.readdir dir |> Array.to_list in
+  List.iter
+    (fun f ->
+      let open Fpath in
+      let source_file = append (to_dir_path (v dir)) (v f) in
+      gen_rule_for_source_file (to_string source_file) (v f))
+    filenames;
+  let filenames = List.map (fun f -> Fpath.v f) filenames in
```

Comment:
I’m not sure I understand clearly why you suggest the changes. I think my intuition of what’s already going on is right; but I’m not looking for reasons to preserve it. I’m happy to make changes having understood clearly what you mean

Reply:
With the change I suggested, paths will be converted to `Fpath` sooner and the raw string arguments will never be used. Also, I suggest changing the `List.iter` code above to take fpaths instead of strings.

Reply:
Done!